test-data/unit/fine-grained-dataclasses.test

```diff
@@ -0,0 +1,39 @@
+-- Test cases for fine-grained incremental checking of dataclasses
+--
+-- The comment at the top of fine-grained.test explains how these tests
+-- work.
+
+[case testDataclassesAsdictFineGrained]
+# flags: --python-version 3.8
+[file a.py]
+from dataclasses import dataclass
+from b import AttributeInOtherModule
+
+@dataclass
+class MyDataclass:
+    attr: AttributeInOtherModule
+
+[file b.py]
+AttributeInOtherModule = str
+[file c.py]
+from dataclasses import asdict
+from a import MyDataclass
```

Comment:
Don't import the class itself, import a variable with this type from `a`, say `my_var: MyDataclass`.